File: DSA-CPP/RemoveDuplicateSortedArray.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

class RemoveDuplicateSortedArray {

public:
    // Brute force solution
    // Time complexity: O(NlogN + N)
    static void solution1(vector<int>& arr) {
        set<int> uniqueElements;

        // Add elements to set (set automatically removes duplicates)
        for (int num : arr) {
            uniqueElements.insert(num);
        }

        // Convert set to vector and sort it
        vector<int> uniqueArray(uniqueElements.begin(), uniqueElements.end());
        sort(uniqueArray.begin(), uniqueArray.end());

        cout << "Brute force solution:" << endl;
        for (int i = 0; i < uniqueArray.size(); i++) {
            cout << uniqueArray[i] << " ";
        }
        cout << endl;
    }

    // Optimal solution
    // Time complexity: O(N)
    static void solution2(vector<int>& arr) {
        int index = 0;

        for (int j = 1; j < arr.size(); j++) {
            if (arr[index] != arr[j]) {
                index++;
                arr[index] = arr[j];
            }
        }

        cout << "Optimal solution:" << endl;
        for (int i = 0; i <= index; i++) {
            cout << arr[i] << " ";
        }
        cout << endl;
    }

    // Python way - Using set to remove duplicates
    static void solution3(vector<int>& arr) {
        set<int> uniqueSet(arr.begin(), arr.end());
        cout << "Python way - ";
        for (int num : uniqueSet) {
            cout << num << " ";
        }
        cout << endl;
    }
};

int main() {
    vector<int> arr = {1, 1, 2, 2, 2, 3, 3};

    // Make a copy of the array for solution1
    vector<int> arr1 = arr;
    RemoveDuplicateSortedArray::solution1(arr1);

    // Solution 2 works in-place, so we use the original array
    RemoveDuplicateSortedArray::solution2(arr);

    // Solution 3
    RemoveDuplicateSortedArray::solution3(arr);

    return 0;
}

File: Python-DSA/RemoveDuplicateSortedArray.py
"""

    Remove duplicates from sorted array

"""

class RemoveDuplicateSortedArray:
    # Brute force solution
    # Time complexity: NlogN + N
    @staticmethod
    def solution1(arr):
        unique_elements = set(arr)

        arr[:len(unique_elements)] = sorted(unique_elements)

        print("Brute force solution:")
        for i in range(len(unique_elements)):
            print(arr[i], end=" ")
        print()

    # Optimal solution
    # Time complexity: O(N)
    @staticmethod
    def solution2(arr):
        index = 0
        for j in range(1, len(arr)):
            if arr[index] != arr[j]:
                index += 1
                arr[index] = arr[j]

        print("Optimal solution:")
        for i in range(index + 1):
            print(arr[i], end=" ")
        print()

    @staticmethod
    def solution3(arr):
        unique_set = set(arr)
        unique_list = list(unique_set)
        print(f"python way - {unique_list}")

if __name__ == "__main__":
    arr = [1, 1, 2, 2, 2, 3, 3]

    arr1 = arr.copy()
    RemoveDuplicateSortedArray.solution1(arr1)
    RemoveDuplicateSortedArray.solution2(arr)
    RemoveDuplicateSortedArray.solution3(arr1)
